packages/supermassive/src/jsutils/PromiseOrValue.ts

```diff
@@ -1 +1,25 @@
-export type PromiseOrValue<T> = Promise<T> | T;
+type ImmutablePrimitive =
+  | undefined
+  | null
+  | boolean
+  | string
+  | number
+  | Function
+  | unknown;
+
+export type Immutable<T> = T extends ImmutablePrimitive
+  ? T
+  : T extends Array<infer U>
+  ? ImmutableArray<U>
+  : T extends Map<infer K, infer V>
+  ? ImmutableMap<K, V>
+  : T extends Set<infer M>
+  ? ImmutableSet<M>
+  : ImmutableObject<T>;
+
+export type ImmutableArray<T> = ReadonlyArray<Immutable<T>>;
+export type ImmutableMap<K, V> = ReadonlyMap<Immutable<K>, Immutable<V>>;
+export type ImmutableSet<T> = ReadonlySet<Immutable<T>>;
+export type ImmutableObject<T> = { readonly [K in keyof T]: Immutable<T[K]> };
+
+export type PromiseOrValue<T> = Immutable<T> | Promise<Immutable<T>>;
```

Comment:
I have to wonder how expensive this might be vs the previous inlined solution, when compiling or trying to get instant feedback from the LSP.

Reply:
Yeah, there are other problems with it as well (posted in a separate comment). Let's stick with a simple approach then (at least for now)